sre-recipes/recipe_runner.py

```diff
@@ -0,0 +1,264 @@
+
+    def run_break(self):
+        print('Deploying broken service...')
+        for action in self.config.get("break", []):
+            self.action_handler.handle_action(action)
+        print('Done. Deployed broken service')
+
+    def run_restore(self):
+        print('Restoring broken service...')
+        for action in self.config.get("restore", []):
+            self.action_handler.handle_action(action)
+        print('Done. Restored broken service to working state.')
+
+    def run_hint(self):
+        hint = self.config.get("hint", None)
+        if hint:
+            print(f'Here is your hint!\n\n{hint}')
+        else:
+            print("This recipe has no hints.")
+
+    def run_verify(self):
+        verify_config = self.config.get("verify", [])
+        if not verify_config:
+            raise NotImplementedError("Verify is not configured")
+        for action in verify_config:
+            self.action_handler.handle_action(action)
+
+
+class ActionHandler:
+    """A utility helper for executing actions supported by SRE Recipe configs.
+
+    Implementation Guide
+    --------------------
+    1. Map the action name to the action handler in the `__init__` method.
+    2. All action handlers should take exactly one argument, which is the full 
+       config specified for the action itself, as it is defined in YAML.
+       For example: {action: "run-shell-commands", commands: ['echo Hi']}
+
+    This runner will propgate all exceptions to the caller, and it is caller's
+    responsibility to handle any exception and to perform any error logging.
+    """
+
+    def __init__(self):
+        # Action types to action handlers
+        self.action_map = {
+            "run-shell-commands": self.run_shell_commands,
+            "multiple-choice-quiz": self.run_multiple_choice_quiz,
+            "loadgen-spawn": self.loadgen_spawn,
+            "loadgen-stop": self.loadgen_stop,
+        }
+
+        # Reusable parameters shared between action handlers
+        self.loadgen_ip = None
+
+    def handle_action(self, config):
+        if "action" not in config:
+            raise ValueError("Action config missing `action` type")
+        action_type = config["action"]
+        if action_type not in self.action_map:
+            raise NotImplementedError(
+                f"Action type not implemented: {action_type}")
+        return self.action_map[action_type](config)
+
+    def init_loadgen_ip(self):
+        if not self.loadgen_ip:
+            self.loadgen_ip, err = utils.get_loadgen_ip()
+            if err:
+                raise RuntimeError(f"Failed to get loadgen IP: {err}")
+
+    ############################ Action Handlers ###############################
+
+    def run_shell_commands(self, config):
+        """Runs the commands one at a time in shell.
+
+        Config Paramters
+        ----------------
+        commands: string[]
+            Required. A list of shell command strings.
```

Comment:
how the quotes will be handled in this case?
Say we want to execute command that requires passing value with single or double quotes?

Reply:
No special handling is required.